angular/index.d.ts

```diff
@@ -756,7 +756,7 @@ declare namespace angular {
          * @param comparator Function used to determine the relative order of value pairs.
          * @return An array containing the items from the specified collection, ordered by a comparator function based on the values computed using the expression predicate.
          */
-        <T>(array: T[], expression: string|((value: T) => any)|(((value: T) => any)|string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```

Comment:
Oh interesting.  If you turn on `strictNullChecks` in the playground, it starts correctly determining the types instead of making them `any`.  But with the option off, you are correct, this leads to incorrect typing.  Maybe the only way to deal with it in both situations is to list out all 4 signatures. [playground](http://www.typescriptlang.org/play/#src=interface%20OrderBy%20%7B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cnull%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cnull%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cundefined%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cundefined%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cnull%7Cundefined%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cnull%7Cundefined%3B%0D%0A%7D%0D%0A%0D%0Alet%20f%3A%20OrderBy%3B%0D%0Alet%20s%3A%20string%5B%5D%3B%0D%0Alet%20n%3A%20string%5B%5D%20%7C%20null%3B%0D%0Alet%20u%3A%20string%5B%5D%20%7C%20undefined%3B%0D%0Alet%20nu%3A%20string%5B%5D%20%7C%20null%20%7C%20undefined%3B%0D%0A%0D%0Alet%20sr%20%3D%20f(s)%3B%0D%0Alet%20nr%20%3D%20f(n)%3B%0D%0Alet%20ur%20%3D%20f(u)%3B%0D%0Alet%20nur%20%3D%20f(nu)%3B%0D%0A)